trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+from typing import Mapping
+
+import gpflow
+import numpy as np
+import tensorflow as tf
+
+from trieste.data import Dataset
+from trieste.type import TensorType
+
+jitter = gpflow.config.default_jitter()
+
+
+def atleast2d(tensor: tf.Tensor) -> tf.Tensor:
+    if tf.rank(tensor) <= 1:
+        return tf.expand_dims(tensor, axis=0)
+    return tensor
+
+
+def to_Dataset(datasets: Mapping[str, Dataset]) -> Dataset:
+
+    data_query_points = []
+    data_observations = []
+
+    for tag in datasets:
+        data_query_points.append(datasets[tag].query_points)
+        data_observations.append(datasets[tag].observations)
+
+    assert tf.math.reduce_all(
+        [tf.math.equal(input_data, data_query_points[0]) for input_data in data_query_points]
+    ), ValueError("Input data is not the same for each objective")
+
+    mo_dataset = Dataset(data_query_points[0], tf.concat(data_observations, axis=1))
+
+    return mo_dataset
+
+
+def non_dominated_sort(datasets: Dataset) -> TensorType:
+    """
+    Computes the non-dominated set for a set of data points
+    :param dataset: data points
+    :return: tuple of the non-dominated set and the degree of dominance,
+        dominances gives the number of dominating points for each data point
+    """
+    observations = datasets.observations
+    extended = tf.tile(tf.expand_dims(observations, 0), [observations.shape[0], 1, 1])
+    swapped_ext = tf.einsum("ij...->ji...", extended)
```

Comment:
this assumes `observations` has rank 2, so to avoid einsum, we can also write
```
tf.transpose(tf.tile(xs[None, ...], [len(xs), 1, 1]), [1, 0, 2])
```
or even
```
tf.reshape(tf.tile(xs, [1, len(xs)]), [len(xs), len(xs), -1])
```